yarn-project/noir-libs/value-note/src/utils.nr

```diff
@@ -58,6 +60,105 @@ fn spend_notes(
     );
 }
 
+/*
+  Spends one note from a set of 4 notes.
+  Details: Reads 4 notes from the user's `balance`: [n_{o}, n_{o+1}, n_{o+2}, n_{o+3}]
+  where "o" is the note_offset. Then, spends the note: [n_{o + i}] 
+  where "i" is the spend_note_index. This gives more control to the user on which
+  note is to be spent. 
+*/
+fn spend_one_note(
+    context: &mut Context,
+    balance: Set<ValueNote, VALUE_NOTE_LEN>,
+    amount: Field,
+    owner: Field,
+    note_offset: Field,
+    spend_note_index: Field,
+) {
+    let options = NoteGetterOptions::with_filter_and_offset(get_4_notes, 0, note_offset as u32);
+    let maybe_notes = balance.get_notes(context, options);
+
+    let note0 = maybe_notes[0].unwrap_or(ValueNote::dummy());
+    let note1 = maybe_notes[1].unwrap_or(ValueNote::dummy());
+    let note2 = maybe_notes[2].unwrap_or(ValueNote::dummy());
+    let note3 = maybe_notes[3].unwrap_or(ValueNote::dummy());
+
+    assert(spend_note_index as u64 < 4);
+
+    // Ensure the notes are actually owned by the owner (to prevent user from generating a valid proof while
+    // spending someone else's notes).
+    if maybe_notes[0].is_some() {
```

Comment:
so save number of gates, maybe only assert for the note whose index we want to spend?

Reply:
Right, good point.